### ISSUE TITLE
TimeSeries: very high memory usage on web browsers

### ISSUE THREAD

**Issue by xpduyson** [grafana/grafana#35987]
**What happened**:
I'm using an arm64 client with 2GB of RAM as an Chromium kiosk dashboard. Adding TimeSeries chart to the dashboard make Chromium memory usage keep increasing to the point it completely unresponsive. A dashboard with 3 TimeSeries chart would consume up to 1.3GB of memory in 15 minutes.

**What you expected to happen**:
Memory leak should not happened.

**How to reproduce it (as minimally and precisely as possible)**:
- Create a dashboard with TimeSeries charts. Add many Time Series charts will make the memory leaks happen faster.
- Open it with Chromium and wait till it not responding anymore.
- Please see the attached video (5x speed up) for demonstration:

https://user-images.githubusercontent.com/6981890/122753428-7398a400-d2bc-11eb-813c-317c81c203e6.mp4

**Anything else we need to know?**:

**Environment**:
- Grafana version: 8.0.3
- Data source type & version: Prometheus 2.27.1
- OS Grafana is installed on: Debian Buster
- User OS & Browser: Debian Buster, Chromium 90
- Grafana plugins:
- Others:

**Comment by torkelo** [comment]
Thanks for reporting this! can confirm we have a big issue here, we will investigate and get this fixed asap

**Comment by perplexityjeff** [comment]
Thanks for this issue, we can confirm that this is an issue as well in our Kiosk systems but tested this on Firefox as well. Firefox is more efficient with memory usage but it does the same thing. 

We tested on Raspberry Pi's and on Windows machines as well.

**Comment by dprokop** [comment]
@xpduyson @perplexityjeff thanks for bringing this up. I have done some investigation with the following findings. I can confirm the problem is present in 8.0.3 - this can be easily reproduced against dashboard on play.grafana.com[1]. 

In recent days we [had an issue reported](https://github.com/grafana/grafana/issues/35848) about uPlot(TimeSeries panel's underlying library) being unnecessarily reinitialized on every data update. I have used that issue as a good reference for investigating the memory consumption of TimeSeries panel.

I have collected performance profiles on play[1] and locally against the attached dashboard[2]. Locally, the profile was collected against the main branch after the fix[3] (e94b82c833d09dbecd5948d764fe8f801446b351) and before the fix[3] applied (main with 6f3887ca884f06d56e3d099d5902ff61b6b11ecb reverted).

The profiles were collected agains exactly the same dashboard over 2min period of time.

Profiles collected on main with the fix reverted and on play[1] show consitent memory usage increase:
1. play.grafana.com[1]: 
<img width="1298" alt="Screen Shot 2021-06-22 at 13 14 18" src="https://user-images.githubusercontent.com/2376619/122916330-0d179280-d35d-11eb-9165-686ce23b9c92.png">
2. localhost (main with 6f3887ca884f06d56e3d099d5902ff61b6b11ecb reverted)
<img width="1298" alt="Screen Shot 2021-06-22 at 13 14 27" src="https://user-images.githubusercontent.com/2376619/122916372-19035480-d35d-11eb-9a16-b7f10405aa57.png">

The same profile collected agains latest main shows a completely different situation:
<img width="1298" alt="Screen Shot 2021-06-22 at 13 14 33" src="https://user-images.githubusercontent.com/2376619/122916617-5f58b380-d35d-11eb-86be-c7adba260cfb.png">

It seems like the GC kicks in more effectively, yet you can still notice there are some leftover listeners not being removed over the time. This can be noticed but comparing the graph of listeners in the document. The number of listeners increases a bit after the reinitialization. While the fix[3] mitigates the problem to some extend, I am afraid the underlying code can still lead to memory leaks, but over a way longer period of time. I have not identified the real underlying issue yet, but Grafana 8.0.4 coming out soon should significantly improve your experience with TimeSeries panel.

I'll keep you updated on the findings.


[1] https://play.grafana.org/d/9hY_5lR7z/memory-leak-ts?orgId=1&refresh=5s
[2] 
```
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "gnetId": null,
  "graphTooltip": 0,
  "id": 487,
  "links": [],
  "panels": [
    {
      "datasource": "gdev-testdata",
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 2,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "single"
        }
      },
      "title": "Panel Title",
      "type": "timeseries"
    },
    {
      "datasource": "gdev-testdata",
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 8,
        "x": 12,
        "y": 0
      },
      "id": 11,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "single"
        }
      },
      "title": "Panel Title",
      "type": "timeseries"
    },
    {
      "datasource": "gdev-testdata",
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 9
      },
      "id": 10,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "single"
        }
      },
      "title": "Panel Title",
      "type": "timeseries"
    },
    {
      "datasource": "gdev-testdata",
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 8,
        "w": 6,
        "x": 12,
        "y": 9
      },
      "id": 9,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "single"
        }
      },
      "title": "Panel Title",
      "type": "timeseries"
    },
    {
      "datasource": "gdev-testdata",
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 12,
        "y": 17
      },
      "id": 7,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "single"
        }
      },
      "title": "Panel Title",
      "type": "timeseries"
    },
    {
      "datasource": "gdev-testdata",
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 18
      },
      "id": 8,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "single"
        }
      },
      "title": "Panel Title",
      "type": "timeseries"
    },
    {
      "datasource": "gdev-testdata",
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 12,
        "y": 26
      },
      "id": 5,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "single"
        }
      },
      "title": "Panel Title",
      "type": "timeseries"
    },
    {
      "datasource": "gdev-testdata",
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 27
      },
      "id": 6,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "single"
        }
      },
      "title": "Panel Title",
      "type": "timeseries"
    },
    {
      "datasource": "gdev-testdata",
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 12,
        "y": 35
      },
      "id": 3,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "single"
        }
      },
      "title": "Panel Title",
      "type": "timeseries"
    },
    {
      "datasource": "gdev-testdata",
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 36
      },
      "id": 4,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "single"
        }
      },
      "title": "Panel Title",
      "type": "timeseries"
    }
  ],
  "refresh": "5s",
  "schemaVersion": 30,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": []
  },
  "time": {
    "from": "2021-06-22T06:17:29.024Z",
    "to": "2021-06-22T07:30:21.678Z"
  },
  "timepicker": {},
  "timezone": "",
  "title": "Memory leak TS",
  "uid": "CWRMt_Rnz",
  "version": 1
}
```
[3] https://github.com/grafana/grafana/pull/35952

**Comment by dprokop** [comment]
Did some more investigation. I'm particularly interested in memory consumption of the time series panel, so as a baseline I took a simple dashboard with multiple Text panel to see how Grafana's memory consumption changes over the same period of time for simple vs complex panel. The recording periods are 2min. All recordings performed on a dashbaord with 5s refresh interval

1. Profiling dashboard- multiple text panels[1]-. We can see a consistent number of listeners being added on every refresh. At the same time, the memory usage is increasing consistently, but we can see GC kicking in.
<img width="1298" alt="Screen Shot 2021-06-22 at 14 59 18_baseline" src="https://user-images.githubusercontent.com/2376619/122928813-810c6780-d36a-11eb-8aee-23e6104d43e6.png">


2. Time series panel with fix[2] applied. As we can see the listeners added by time series panel are all cleared on every refresh. The memory consumption increases consistently with GC **not** kicking in
<img width="1298" alt="Screen Shot 2021-06-22 at 14 35 31-basemain" src="https://user-images.githubusercontent.com/2376619/122926890-9aacaf80-d368-11eb-8ca1-8e1bbafe25c8.png">

3. Time series panel rendering empty div. We can see consistent number of listeners being added on every refresh. GC kicks in now end then releasing some of the memory. Yet - the memory is consistently increasing.
  <img width="1298" alt="Screen Shot 2021-06-22 at 14 06 21-timeseries_disabled" src="https://user-images.githubusercontent.com/2376619/122926744-7a7cf080-d368-11eb-81e4-c02c108cea65.png">

Profile 1 and 3 are quite similar - in both cases the memory usage is increasing on a similar level (Δ ~2MB)
The difference between 2. and 3. is that there is no UPlotSeriesBuilder included in the process, yet the memory is still increasing (Δ ~ 3MB,( allowing for some GC intervention in 3.). Wonder if the issue isn't somewhere in https://github.com/grafana/grafana/blob/60f5865ee2c59dbde8b493bace71cdf0c06f8a00/public/app/plugins/panel/timeseries/utils.ts#L13 as it creates a new frame every time data changes?

[1] 
```
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "gnetId": null,
  "graphTooltip": 0,
  "id": 489,
  "links": [],
  "panels": [
    {
      "datasource": null,
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 2,
      "options": {
        "content": "# Title\n\n  For markdown syntax help: [commonmark.org/help](https://commonmark.org/help/)\n           ",
        "mode": "markdown"
      },
      "pluginVersion": "8.1.0-pre",
      "title": "Panel Title",
      "type": "text"
    },
    {
      "datasource": null,
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 12,
        "y": 0
      },
      "id": 7,
      "options": {
        "content": "# Title\n\n  For markdown syntax help: [commonmark.org/help](https://commonmark.org/help/)\n           ",
        "mode": "markdown"
      },
      "pluginVersion": "8.1.0-pre",
      "title": "Panel Title",
      "type": "text"
    },
    {
      "datasource": null,
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 9
      },
      "id": 6,
      "options": {
        "content": "# Title\n\n  For markdown syntax help: [commonmark.org/help](https://commonmark.org/help/)\n           ",
        "mode": "markdown"
      },
      "pluginVersion": "8.1.0-pre",
      "title": "Panel Title",
      "type": "text"
    },
    {
      "datasource": null,
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 12,
        "y": 9
      },
      "id": 5,
      "options": {
        "content": "# Title\n\n  For markdown syntax help: [commonmark.org/help](https://commonmark.org/help/)\n           ",
        "mode": "markdown"
      },
      "pluginVersion": "8.1.0-pre",
      "title": "Panel Title",
      "type": "text"
    },
    {
      "datasource": null,
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 18
      },
      "id": 4,
      "options": {
        "content": "# Title\n\n  For markdown syntax help: [commonmark.org/help](https://commonmark.org/help/)\n           ",
        "mode": "markdown"
      },
      "pluginVersion": "8.1.0-pre",
      "title": "Panel Title",
      "type": "text"
    },
    {
      "datasource": null,
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 12,
        "y": 18
      },
      "id": 9,
      "options": {
        "content": "# Title\n\n  For markdown syntax help: [commonmark.org/help](https://commonmark.org/help/)\n           ",
        "mode": "markdown"
      },
      "pluginVersion": "8.1.0-pre",
      "title": "Panel Title",
      "type": "text"
    },
    {
      "datasource": null,
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 27
      },
      "id": 8,
      "options": {
        "content": "# Title\n\n  For markdown syntax help: [commonmark.org/help](https://commonmark.org/help/)\n           ",
        "mode": "markdown"
      },
      "pluginVersion": "8.1.0-pre",
      "title": "Panel Title",
      "type": "text"
    },
    {
      "datasource": null,
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 12,
        "y": 27
      },
      "id": 3,
      "options": {
        "content": "# Title\n\n  For markdown syntax help: [commonmark.org/help](https://commonmark.org/help/)\n           ",
        "mode": "markdown"
      },
      "pluginVersion": "8.1.0-pre",
      "title": "Panel Title",
      "type": "text"
    }
  ],
  "refresh": "5s",
  "schemaVersion": 30,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": []
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "Memory leak soft",
  "uid": "NqdXa9R7z",
  "version": 1
}
```
[2] #35952

**Comment by perplexityjeff** [comment]
@dprokop Any update regarding this issue? Currently we have to restart our various kiosk systems periodically to prevent issues.

**Comment by dprokop** [comment]
@perplexityjeff yup, 8.0.4 coming out on Thursday this week and it will help with the high memory usage.

**Comment by leeoniya** [comment]
interestingly, my listener count is flat, but the heap does grow after each manual data refresh. i did a forced GC at the end and it had little effect. this is 6 panels 2 random walk series each:

![image](https://user-images.githubusercontent.com/43234/123878676-3019e780-d905-11eb-96e7-432226710d29.png)

**Comment by leeoniya** [comment]
the growth in listeners only happens in dev builds, but they do get collected periodically:

![image](https://user-images.githubusercontent.com/43234/123880444-48d7cc80-d908-11eb-917b-9de422b7e9fc.png)

**Comment by leeoniya** [comment]
the leak is happening somewhere in here:

https://github.com/grafana/grafana/blob/8aeb1cc030a72f647a6dfa9acdbf62e2d8aa9e44/public/app/features/query/state/PanelQueryRunner.ts#L258-L262

**Comment by xpduyson** [comment]
Upgraded to v8.0.4 last night and the memory usage situation has significantly improved. I'm going to disable the restart and refresh timer of the kiosk to see how long it can survive.

![Screenshot from 2021-07-02 09-46-19](https://user-images.githubusercontent.com/6981890/124215019-72b60c80-db1d-11eb-947d-58569f3fcb27.jpg)

**Comment by perplexityjeff** [comment]
As @xpduyson, we are as well going to test the 8.0.4 improvements, Let's hope for the best.

**Comment by leeoniya** [comment]
the PR linked above (will be in 8.0.5) fixes the remaining leak in my tests.

before PR:

![image](https://user-images.githubusercontent.com/43234/124281611-80fd2c00-db0f-11eb-9e9e-66f1302fe490.png)

after PR:

![image](https://user-images.githubusercontent.com/43234/124281731-a2f6ae80-db0f-11eb-8342-7e9cc5bd2c51.png)

**Comment by xpduyson** [comment]
With v8.0.4 and Chromium, a dashboard with 3 Time series charts slowly consume up to 1GB of RAM in around 26h. So a 24h restart timer workaround is good enough for me to wait till v8.0.5. 

![Screenshot from 2021-07-03 15-56-10](https://user-images.githubusercontent.com/6981890/124349682-bd697e80-dc1a-11eb-9d24-949b6a00d9c1.png)

**Comment by qiujiangkun** [comment]
I have the same issue, from 6.x to 8.x. My memory consumption goes up to 5 GiB after several hours' idle.

**Comment by tskarhed** [comment]
This seems to have been fixed in #36371. The fix will be out in v8.0.5.
